Mathlib/Analysis/Fourier/ZMod.lean

```diff
@@ -0,0 +1,125 @@
+  notation `𝓕`, scoped in namespace `ZMod`, is available for this.
+* `DirichletCharacter.fourierTransform_eq_inv_mul_gaussSum`: the discrete Fourier transform of a
+  primitive Dirichlet character `χ` is a Gauss sum times `χ⁻¹`.
+-/
+
+open scoped Real
+
+namespace ZMod
+
+/-- The `AddMonoidHom` from `ZMod N` to `ℝ / ℤ` sending `j mod N` to `j / N mod 1`. -/
+noncomputable def toAddCircle {N : ℕ+} : ZMod N →+ UnitAddCircle :=
+  lift N ⟨AddMonoidHom.mk' (fun j ↦ ↑(j / N : ℝ)) (by simp [add_div]), by simp⟩
+
+lemma toAddCircle_coe {N : ℕ+} (j : ℤ) :
+    toAddCircle (j : ZMod N) = ↑(j / N : ℝ) := by
+  simp [toAddCircle]
+
+lemma toAddCircle_apply {N : ℕ+} (j : ZMod N) :
+    toAddCircle j = ↑(j.val / N : ℝ) := by
+  conv_lhs => rw [show j = (val j : ℤ) by simp, toAddCircle_coe]
+  simp only [natCast_val, intCast_cast]
+
+lemma toAddCircle_injective (N : ℕ+) : Function.Injective (toAddCircle : ZMod N → _) := by
+  intro x y hxy
+  have : (0 : ℝ) < N := Nat.cast_pos.mpr N.pos
+  rwa [toAddCircle_apply, toAddCircle_apply, AddCircle.coe_eq_coe_iff_of_mem_Ico
+    (hp := Real.fact_zero_lt_one) (a := 0), div_left_inj' this.ne', Nat.cast_inj,
+    (val_injective N).eq_iff] at hxy <;>
+  exact ⟨by positivity, by simpa only [zero_add, div_lt_one this, Nat.cast_lt] using val_lt _⟩
+
+/-- The additive character from `ZMod N` to the unit circle in `ℂ`, sending `j mod N` to
+`exp (2 * π * I * j / N)`. -/
+noncomputable def toCircle {N : ℕ+} : AddChar (ZMod N) circle where
+  toFun := fun j ↦ (toAddCircle j).toCircle
+  map_add_mul' a b := by simp_rw [map_add, AddCircle.toCircle_add]
+  map_zero_one' := by simp_rw [map_zero, AddCircle.toCircle, ← QuotientAddGroup.mk_zero,
+    Function.Periodic.lift_coe, mul_zero, expMapCircle_zero]
+
+lemma toCircle_coe {N : ℕ+} (j : ℤ) :
+    toCircle (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  rw [toCircle, AddChar.coe_mk, AddCircle.toCircle, toAddCircle_coe,
+    Function.Periodic.lift_coe, expMapCircle_apply]
+  push_cast
+  ring_nf
+
+lemma toCircle_apply {N : ℕ+} (j : ZMod N) :
+    toCircle j = Complex.exp (2 * π * Complex.I * j.val / N) := by
+  rw [← Int.cast_natCast, ← toCircle_coe, natCast_val, intCast_zmod_cast]
+
+/-- The additive character from `ZMod N` to `ℂ`, sending
+`j mod N` to `exp (2 * π * I * j / N)`. -/
+noncomputable def stdAddChar {N : ℕ+} : AddChar (ZMod N) ℂ :=
+  (Submonoid.subtype circle).compAddChar toCircle
+
+lemma stdAddChar_coe {N : ℕ+} (j : ℤ) :
+    stdAddChar (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  simp only [stdAddChar, MonoidHom.coe_compAddChar, Function.comp_apply,
+    Submonoid.coe_subtype, toCircle_coe]
+
+lemma stdAddChar_apply {N : ℕ+} (j : ZMod N) :
+    stdAddChar j = ↑(toCircle j) :=
+  rfl
+
+section fourier
+
+open MeasureTheory
+
+/-- The discrete measurable space structure (every set is measurable). -/
+local instance instMeasurableSpaceZMod (N : ℕ+) : MeasurableSpace (ZMod N) := ⊤
```

Comment:
Should this use any `n : Nat` together with an assumption `[NeZero n]`? It would apply to `ZMod n` even if `n` a `PNat` coerced to `Nat`.

Reply:
Aha, that's a really good idea! In the next phase of the project I was constantly getting tangled in things like (1 : ℕ+) not being the same as `(1 : ℕ)` and I was sure there had to be a better way. This is clearly the right approach, thanks!